src/algoseek_connector/base.py

```diff
@@ -0,0 +1,1042 @@
+        self._source = source
+        self._description = description
+        self.metadata = MetaData()
+        datasets = [DataSetFetcher(self, x) for x in self.list_datasets()]
+        self.datasets = DataSetMapping(*datasets)
+
+    @property
+    def description(self) -> DataGroupDescription:
+        """Get the data group description."""
+        return self._description
+
+    @property
+    def source(self) -> DataSource:
+        """Get the data source."""
+        return self._source
+
+    def fetch_dataset(self, name: str) -> DataSet:
+        """
+        Load a dataset from a data source.
+
+        Parameters
+        ----------
+        name : str
+            The dataset name.
+
+        Raises
+        ------
+        InvalidDataSetName
+            If an invalid dataset name is provided.
+
+        """
+        return self.datasets[name].fetch()
+
+    def list_datasets(self) -> list[str]:
+        """List available datasets."""
+        return self._source.client.list_datasets(self.description.name)
+
+    def _repr_html_(self):  # pragma: no cover
+        """Display the DataGroup in jupyter notebooks using HTML."""
+        return self.description.html()
+
+
+class DataSetMapping:
+    """
+    Mapping class that stores Datasets from a DataGroup.
+
+    Implements the Mapping protocol.
+    """
+
+    def __init__(self, *datasets: DataSetFetcher):
+        for ds in datasets:
+            setattr(self, ds.description.name, ds)
+
+    def __len__(self):
+        return len(self.__dict__)
+
+    def __iter__(self) -> Generator[str, None, None]:
+        yield from self.__dict__
+
+    def __getitem__(self, key: str) -> DataSetFetcher:
+        try:
+            return self.__dict__[key]
+        except KeyError:
+            raise InvalidDataSetName(key) from None
+
+    def _ipython_key_completions_(self):  # pragma: no cover
+        return list(self.__dict__)
+
+
+class DataSetFetcher:
+    """
+    Lightweight representation of a dataset.
+
+    Manages creation of DataSet instances for querying data using SQL and
+    data downloading.
+
+    Methods
+    -------
+    download:
+        Download data files.
+    fetch:
+        Create a DataSet instance.
+
+    """
+
+    def __init__(self, group: DataGroup, name: str):
+        self._group = group
+        self._source = group.source
+        group_name = self.group.description.name
+        self._description = group.source.description_provider.get_dataset_description(
+            group_name, name
+        )
+        self._dataset = None
+
+    @property
+    def description(self) -> DataSetDescription:
+        """Get the dataset name."""
+        return self._description
+
+    @property
+    def group(self) -> DataGroup:
+        """Get the dataset group."""
+        return self._group
+
+    @property
+    def source(self) -> DataSource:
+        """Get the data source."""
+        return self._source
+
+    def download(
```

Comment:
Currently it is implemented only for S3 datasets, but it may be implemented for ArdaDB if we know the column names where symbol, dates and expiration date information is stored.